### ISSUE TITLE
error: redeclaration of 'foo' cannot add 'dllexport' attribute

### ISSUE THREAD

**Issue by lhmouse** [llvm/llvm-project#55329]
If a function is declared without `dllexport` and called, but is later defined with `dllexport`, clang generates an error. GCC does not suffer from this issue (the attributes are merged, as if all declarations had `dllexport`).

```c
void
foo(void);

void
bar(void)
  {
    foo();  /* the error goes away if this call is deleted  */
  }

__declspec(dllexport)
void
foo(void)  /* test.c:3:28: error: redeclaration of 'foo' cannot add 'dllexport' attribute */
  {
  }
```

```text
E:\lh_mouse\Desktop>clang --version
clang version 14.0.0
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: C:/MSYS2/clang64/bin
```

**Comment by MaskRay** [comment]
This is similar to
https://maskray.me/blog/2021-10-10-when-can-glibc-be-built-with-clang#asm-label-after-first-use

IIRC due to the Clang code gen architecture it is very difficult to emulate GCC behavior.

**Comment by lhmouse** [comment]
Note it is suggested that `dllimport` should not be used at all. When building a DLL, function definitions can be marked `dllexport` as needed, which should not conflict with undecorated declarations. This approach enables the same function declarations (in headers) to be shared between static and shared libraries.

If Clang does not work this way, it is suddenly mandatory to mark function declarations in headers with some nastily complex macros, which also requires a project-specific macro that is invisible to users, as it is not enough to check for just `DLL_EXPORT`; when a user is building their own DLL, they should never see `dllexport` from another library.

**Comment by mstorsjo** [comment]
Yes, I think this error is kinda overzealous - adding a `dllimport` after first being declared without it would be an issue, but adding `dllexport` later should, in principle, not be a problem. The problem is just with Clang and LLVM code generation - which might not be entirely straightforward to fix. As far as I understand the issue, in the example case, the function `foo` doesn't get a declaration in LLVM IR until it's either called or defined, and once declared in the LLVM IR, we can't add the attribute afterwards.

So to fix this, the code generation would need to go back and adjust the already-emitted declaration of `foo` to add the dllexport attribute (on the IR level). Going back and changing attributes for already-emitted IR declarations generally probably isn't supported much (because it's generally not correct to do so), but in the case of a late addition of `dllexport`, it could be ok.

**Comment by mstorsjo** [comment]
@rnk @zmodem (If you happen to have time) - I presume you have more experience with the Clang codegen than me... What would it take to fix the late-addition of dllexport in this case?

**Comment by zmodem** [comment]
The problem is indeed that going back and changing a symbol after it's already been emitted to LLVM IR isn't something Clang codegen normally does. It may of course be possible to do in theory, but I worry that it might be tricky to get completely right.

Since it would be a break with Clang's codegen architecture, I don't think we should try to do this without good motivation, such as widely used system headers relying on it. In general, well-behaved code should be consistent in how it declares its symbols.

**Comment by lhmouse** [comment]
> Since it would be a break with Clang's codegen architecture, I don't think we should try to do this without good motivation, such as widely used system headers relying on it. In general, well-behaved code should be consistent in how it declares its symbols.

How do you define 'consistent'? Suppose:

#### fahr.h

```c
inline double fahr(double t)  // C99 inline definition
{
return (9.0 * t) / 5.0 + 32.0;
}
```

#### fahr.c

```c99
#include "fahr.h"

extern double fahr(double); // creates an external definition
```

This is formally specified by the C99 standard (see _6.7.4 Function specifiers, WG14 N1256_), and how to you expect it to be 'consistent'? The two declarations can never match each other.

**Comment by rnk** [comment]
Perhaps upgrading to dllexport is reasonably safe, it just takes extra work, and we didn't think that work was necessary because [MSVC rejects this example](https://gcc.godbolt.org/z/WqGe4WKh3). GCC compatibility, historically, was not our goal and was improved later.

Clang does have this same problem for C99 inline function definitions, and I seem to recall that it requires a special codepath to go back and upgrade weak inline function definitions into strong definitions. I don't think there are any risks to adding a similar codepath for dllexport, it just doesn't exist yet, and from a general engineering design sense, less code is better.

Note that going the other direction, adding dllimport to a previous declaration, would be much riskier, since it can make the address of a symbol non-constexpr. I am aware of weird cases where clang will revise dllimport functions to dllexport functions, but again, that's moving from weaker linkages to stronger linkages.

**Comment by zmodem** [comment]
@lhmouse can you elaborate on your use case for this? Is there some project that relies on GCC's behavior?

**Comment by lhmouse** [comment]
The example was reduced from here: https://github.com/lhmouse/mcfgthread/blob/61211f574fd65ffb8b056d697539b5734248e536/src/mutex.c#L37

Previous I used `-export-all-symbols`, with `--exclude-symbols` to exclude the custom entry point, so there was no `dllexport`, and hence no mismatching declaration, either. Clang does not support `--exclude-symbols` however, and removal of it would export the entry point function undesirably. So I removed `--export-all-symbols` and decided to mark public functions with `dllexport` when building the DLL.

**Comment by mstorsjo** [comment]
> Previous I used `-export-all-symbols`, with `--exclude-symbols` to exclude the custom entry point, so there was no `dllexport`, and hence no mismatching declaration, either. Clang does not support `--exclude-symbols` however, and removal of it would export the entry point function undesirably.

I presume that would be LLD not supporting `--exclude-symbols` (you could be using Clang with ld.bfd too, where it'd work, or GCC with LLD). Implementing that should be fairly straightforward (I just never bothered as I hadn't run into any use of it in the wild back when I initially added support for exporting all symbols in LLD). Although that doesn't change the fact that it would be nice if dllexport could be added later too.

**Comment by lhmouse** [comment]
> I presume that would be LLD not supporting `--exclude-symbols` (you could be using Clang with ld.bfd too, where it'd work, or GCC with LLD). Implementing that should be fairly straightforward (I just never bothered as I hadn't run into any use of it in the wild back when I initially added support for exporting all symbols in LLD). Although that doesn't change the fact that it would be nice if dllexport could be added later too.

Gave it a try:

```diff
 lib_libmcfgthread_la_LDFLAGS =  \
   -Wl,--subsystem,windows:6.1,--entry,__MCF_dll_startup@@Z  \
-  -Wl,--exclude-all-symbols,--enable-auto-image-base  \
+  -Wl,--export-all-symbols,--enable-auto-image-base  \
   -Wl,--disable-{runtime-pseudo-reloc,stdcall-fixup}  \
-  -Wc,-nostdlib -no-undefined -version-info @abi_major@:@abi_minor@
+  -Wc,-nostdlib,-fuse-ld=ld -no-undefined -version-info @abi_major@:@abi_minor@
```

It then failed with undefined symbols:

```text
  CCLD     lib/libmcfgthread.la
ld: error: undefined symbol: __MCF_mcopy
>>> referenced by src/memcpy.c:16
>>>               src/.libs/lib_libmcfgthread_la-memcpy.o:(memcpy)

ld: error: undefined symbol: __MCF_mmove
>>> referenced by src/memmove.c:16
>>>               src/.libs/lib_libmcfgthread_la-memmove.o:(memmove)

ld: error: undefined symbol: __MCF_mcomp
>>> referenced by src/memcmp.c:16
>>>               src/.libs/lib_libmcfgthread_la-memcmp.o:(memcmp)

ld: error: undefined symbol: __MCF_mfill
>>> referenced by src/memset.c:16
>>>               src/.libs/lib_libmcfgthread_la-memset.o:(memset)
x86_64-w64-mingw32-clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [Makefile:1300: lib/libmcfgthread.la] Error 1
make[1]: Leaving directory '/d/lh_mouse/GitHub/mcfgthread'
make: *** [Makefile:1045: all] Error 2
```

GCC with lld didn't work either:

```text
libtool: link: x86_64-w64-mingw32-gcc -march=core2 -shared  src/.libs/lib_libmcfgthread_la-xwinnt.o src/.libs/lib_libmcfgthread_la-memcpy.o src/.libs/lib_libmcfgthread_la-memmove.o src/.libs/lib_libmcfgthread_la-memcmp.o src/.libs/lib_libmcfgthread_la-memset.o src/.libs/version.o src/.libs/lib_libmcfgthread_la-startup.o src/.libs/lib_libmcfgthread_la-fwd.o src/.libs/lib_libmcfgthread_la-clock.o src/.libs/lib_libmcfgthread_la-mutex.o src/.libs/lib_libmcfgthread_la-cond.o src/.libs/lib_libmcfgthread_la-once.o src/.libs/lib_libmcfgthread_la-thread.o src/.libs/lib_libmcfgthread_la-cxa.o src/.libs/lib_libmcfgthread_la-dtor_queue.o src/.libs/lib_libmcfgthread_la-exit.o src/.libs/lib_libmcfgthread_la-tls.o src/.libs/lib_libmcfgthread_la-gthr_aux.o src/.libs/lib_libmcfgthread_la-gthr.o   -lkernel32 -lntdll  -march=core2 -g -O2 -Wl,--subsystem -Wl,windows:6.1 -Wl,--entry -Wl,__MCF_dll_startup@@Z -Wl,--export-all-symbols -Wl,--enable-auto-image-base -Wl,--disable-runtime-pseudo-reloc -Wl,--disable-stdcall-fixup -nostdlib -fuse-ld=lld   -o lib/.libs/libmcfgthread-1.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker lib/.libs/libmcfgthread.dll.a
collect2.exe: fatal error: cannot find 'ld'
compilation terminated.
make: *** [Makefile:1300: lib/libmcfgthread.la] Error 1
```

**Comment by lhmouse** [comment]
I managed to find a solution:  https://github.com/lhmouse/mcfgthread/commit/bcad10ac4cab17f180ff616496fbf27ff864250b
Really, I dislike this sort of complexity but there seems to be no better alternatives.

**Comment by Pawday** [comment]
I have similar with this example

```c
void public_function(void);

__declspec(dllexport)
void public_function(void) // ExportTest.c:4:6: warning: redeclaration of 'public_function' should not add 'dllexport' attribute [-Wdll-attribute-on-redeclaration]
{

}
#pragma clang diagnostic pop
```

Why clang treats definition as redeclaration I don't actually know

But I have 2 solutions:

* Ignore warning via compile option

```sh
clang -c ExportTest.c -Wno-dll-attribute-on-redeclaration
```

* Ignore warning via diagnostic pragma

```c
#pragma clang diagnostic push
__declspec(dllexport)
#pragma clang diagnostic ignored "-Wdll-attribute-on-redeclaration"
void public_function(void)
{

}
#pragma clang diagnostic pop
```

**Comment by zmodem** [comment]
> I have similar with this example
> 
> ```c
> void public_function(void);
> 
> __declspec(dllexport)
> void public_function(void) // ExportTest.c:4:6: warning: redeclaration of 'public_function' should not add 'dllexport' attribute [-Wdll-attribute-on-redeclaration]
> {
> 
> }
> #pragma clang diagnostic pop
> ```
> 
> Why clang treats definition as redeclaration I don't actually know
> 
> But I have 2 solutions:
> 
> * Ignore warning via compile option
> 
> ```shell
> clang -c ExportTest.c -Wno-dll-attribute-on-redeclaration
> ```
> 
> * Ignore warning via diagnostic pragma
> 
> ```c
> void public_function(void);
> 
> #pragma clang diagnostic push
> #pragma clang diagnostic ignored "-Wdll-attribute-on-redeclaration"
> 
> __declspec(dllexport)
> void public_function(void)
> {
> 
> }
> 
> #pragma clang diagnostic pop
> ```

The real fix would be to put `__declspec(dllexport)` on the first declaration of `public_function`.

**Comment by zmodem** [comment]
> Requiring `__declspec(dllexport)` would require extra work about copying the `#if...#endif` block into the header.

If you put it in the header, you won't need it in the .c file.

> It is useless to users for 'libfoo'. It exists solely to please Clang. It is not good.

It can be useful for users of `libfoo` to know that the function lives in another dll. The way it's usually done is that the header marks the function as `dllexport` when building the library, and `dllimport` when using the library.

Normally this is done with some macro, such as with `DLL_PUBLIC` in this example from GCC: https://gcc.gnu.org/wiki/Visibility#How_to_use_the_new_C.2B-.2B-_visibility_support

**Comment by lhmouse** [comment]
> If you put it in the header, you won't need it in the .c file.

Yes. But as stated above, we would like the header to be usable with either a static library or a shared library.

> It can be useful for users of `libfoo` to know that the function lives in another dll. The way it's usually done is that the header marks the function as `dllexport` when building the library, and `dllimport` when using the library.

The `dllimport` attribute is not necessary on functions, and is not recommended. Calling a `dllimport` function but linking with its static library will fail. If we took this approach, we would be requiring our users to be clear whether they would like to link against a static or shared library. Not only is it not good, it also requires quite a little work for some tools to define a macro according to whether a pre-installed library is static or shared.

**Comment by mstorsjo** [comment]
I also personally agree that this warning is somewhat redundant. Whether to call functions with dllimport or not, is the users choice (some prefer to call it strictly with dllimport, some prefer to leave it out to allow linking both kinds of libraries without knowing which one it will be).

And for the case when dllexport is set, it doesn't make any difference for other users (within the library) to see the declaration either with or without dllexport.

To me, the only question is how much of an inconvenience it is, within the Clang codebase, to silence the warning specifically for dllexport, without affecting the other cases where the warning is valid.
